lbry/lbry/dht/node.py

```diff
@@ -121,56 +126,55 @@ def stop(self) -> None:
+
+        def set_joined():
+            self.joined.set()
+            log.info(
+                "joined dht, %i peers known in %i buckets", len(self.protocol.routing_table.get_peers()),
+                self.protocol.routing_table.buckets_with_contacts()
+            )
+
         if not self.listening_port:
             await self.start_listening(interface)
         self.protocol.ping_queue.start()
         self._refresh_task = self.loop.create_task(self.refresh_node())
 
-        # resolve the known node urls
-        known_node_addresses = []
-        url_to_addr = {}
-
-        if known_node_urls:
-            for host, port in known_node_urls:
-                address = await resolve_host(host, port, proto='udp')
-                if (address, port) not in known_node_addresses and\
-                        (address, port) != (self.protocol.external_ip, self.protocol.udp_port):
-                    known_node_addresses.append((address, port))
-                    url_to_addr[address] = host
-
-        if known_node_addresses:
-            peers = [
-                make_kademlia_peer(None, address, port)
-                for (address, port) in known_node_addresses
-            ]
-            while True:
-                if not self.protocol.routing_table.get_peers():
-                    if self.joined.is_set():
-                        self.joined.clear()
-                    self.protocol.peer_manager.reset()
-                    self.protocol.ping_queue.enqueue_maybe_ping(*peers, delay=0.0)
-                    peers.extend(await self.peer_search(self.protocol.node_id, shortlist=peers, count=32))
-                    if self.protocol.routing_table.get_peers():
-                        self.joined.set()
-                        log.info(
-                            "Joined DHT, %i peers known in %i buckets", len(self.protocol.routing_table.get_peers()),
-                            self.protocol.routing_table.buckets_with_contacts())
-                    else:
-                        continue
-                await asyncio.sleep(1, loop=self.loop)
-
-        log.info("Joined DHT, %i peers known in %i buckets", len(self.protocol.routing_table.get_peers()),
-                 self.protocol.routing_table.buckets_with_contacts())
-        self.joined.set()
-
-    def start(self, interface: str, known_node_urls: typing.List[typing.Tuple[str, int]]):
-        self._join_task = self.loop.create_task(
-            self.join_network(
-                interface=interface, known_node_urls=known_node_urls
-            )
-        )
+        restored_peers = peers_from_urls(await self._storage.get_peers()) if self._storage else []
+
+        fixed_peers = peers_from_urls([
+            (None, await resolve_host(address, udp_port, 'udp'), udp_port, None)
+            for address, udp_port in known_node_urls or []
+        ])
+
+        seed_peers = restored_peers or fixed_peers
+        fallback = False
+        while seed_peers:
```

Comment:
This can be `while True` and the set_joined outside the loop can be removed - also set_joined won't be duplicated / doesnt have to be a function

Reply:
This way the join task doesn't end immediately if there are no persisted peers and no fixed seeds are found or given.

Reply:
I also don't like the duplication, and I assume it is / was there only because some tests (unnecessarily?) rely on it.

Node with no seeds can still join if another node pings it, correct? Then infinite loop makes sense... (But no fixed peers are worth logging a warning, IMO.)